scala/scala-impl/src/org/jetbrains/plugins/scala/worksheet/ammonite/AmmoniteUtil.scala

```diff
@@ -158,145 +159,104 @@ object AmmoniteUtil {
-            }
+
+        def tryIvy() =
+          firstFileMatchingPattern(s"/$ivyPath", new File(getDefaultCachePath))
+            .find(existsPredicate)
+
+        def tryCoursier() =
+          firstFileMatchingPattern(s"/https/*/*/$mavenPath", new File(getCoursierCachePath))
+            .find(existsPredicate)
+
+        tryIvy() orElse tryCoursier() flatMap { parent =>
+          parent.listFiles().find { cf =>
+            val name = cf.getName
+            prefixPatterns.exists(name.startsWith) &&
+            name.endsWith(".jar") &&
+            !name.endsWith("-sources.jar") &&
+            !name.endsWith("-javadoc.jar")
+          } flatMap { //todo more variants?
+            jarModuleRoot =>
+              Option(
+                JarFileSystem
+                  .getInstance()
+                  .findLocalVirtualFileByPath(jarModuleRoot.getCanonicalPath))
+          }
         }
     }
   }
-  
+
   def isAmmoniteSpecificImport(expr: ScImportExpr): Boolean = {
     val txt = expr.getText
     txt.startsWith(ROOT_EXEC) || txt.startsWith(ROOT_FILE) || txt.startsWith(ROOT_IVY) || txt.startsWith(ROOT_PLUGIN)
   }
-  
-  private def findFileByPattern(pattern: String, predicate: File => Boolean): Option[File] = {
-    abstract class PathPart[T] {
-      protected var current: Option[T] = None
-      def getCurrent: Option[T] = current
-      def hasNext: Boolean = false
-
-      def add(): Boolean
-      def reset()
-    }
-
-    case class SimplePart(p: String) extends PathPart[String] {
-      reset()
-      override def add(): Boolean = {
-        current = None
-        false
-      }
-      override def reset(): Unit = current = Option(p)
-    }
-
-    case class OrPart(ps: Iterable[String]) extends PathPart[String] {
-      private var it = ps.iterator
-      setCurrent()
-
-      override def add(): Boolean = {it.hasNext && {setCurrent(); true} || {current = None; false}}
-      override def hasNext: Boolean = it.hasNext
-      override def reset(): Unit = {
-        it = ps.iterator
-        setCurrent()
-      }
-
-      private def setCurrent() {
-        current = Option(it.next())
-      }
-    }
-
-    case class PathIterator(pathParts: Iterable[PathPart[String]]) extends Iterator[File] {
-      private var it = pathParts.iterator
-      private var currentDigit = pathParts.head
-      private var currentVal: Option[String] = Option(gluePath)
-
-      private def gluePath: String = pathParts.flatMap(_.getCurrent.toList).mkString(File.separator)
 
-      private def advance() {
-        if (!currentDigit.add()) {
-          while (!currentDigit.add() && it.hasNext) currentDigit = it.next()
-          if (currentDigit.getCurrent.isEmpty) return
-          pathParts.takeWhile(_ != currentDigit).foreach(_.reset())
-          currentDigit = pathParts.head
-          it = pathParts.iterator
-        }
-
-        currentVal = Option(gluePath)
-      }
-
-      def hasNext: Boolean = currentVal.isDefined
+  sealed trait FileTree
```

Comment:
I personally found the imperative logic hard to reason with, I replaced it with a simpler chain of functions that build a recursive data structure out of the pattern and use it to derive the list of files the pattern expresses